libpod/runtime_pod_linux.go

```diff
@@ -87,45 +87,58 @@ func (r *Runtime) NewPod(options ...PodCreateOption) (*Pod, error) {
 	if pod.config.UsePodCgroup {
 		logrus.Debugf("Got pod cgroup as %s", pod.state.CgroupPath)
 	}
+	if pod.HasInfraContainer() != pod.SharesNamespaces() {
+		return nil, errors.Errorf("Pods must have an infra container to share namespaces")
+	}
 
 	if err := r.state.AddPod(pod); err != nil {
 		return nil, errors.Wrapf(err, "error adding pod to state")
 	}
 
+	if pod.HasInfraContainer() {
+		ctr, err := r.createInfraContainer(ctx, pod)
+		if err != nil {
+			// Tear down pod, as it is assumed a the pod will contain
+			// a pause container, and it does not.
+			if err2 := r.removePod(ctx, pod, true, true); err2 != nil {
+				logrus.Errorf("Error removing pod after pause container creation failure: %v", err2)
+			}
+			return nil, errors.Wrapf(err, "error adding Infra Container")
+		}
+		pod.state.InfraContainerID = ctr.ID()
+		if err := pod.save(); err != nil {
+			return nil, err
+		}
+	}
+
 	return pod, nil
 }
 
 func (r *Runtime) removePod(ctx context.Context, p *Pod, removeCtrs, force bool) error {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-
-	if !r.valid {
-		return ErrRuntimeStopped
-	}
-
 	if !p.valid {
```

Comment:
Move this block out into RemovePod, in front of the pod lock

Reply:
you mean like this:
```	if !p.valid {
		if ok, _ := r.state.HasPod(p.ID()); !ok {
			// Pod probably already removed
			// Or was never in the runtime to begin with
			return nil
		}
	}


	p.lock.Lock()
	defer p.lock.Unlock()
```
Just making sure because the runtime lock/valid check is in the other order right before...

Reply:
Yes - to match the way we're doing things in `removeContainer()` in runtime.go

Reply:
SGTM